### ISSUE TITLE
Cannot deserialize null collection values

### ISSUE THREAD

**Issue by nmiyake** [palantir/conjure-java#443]
## What happened?
An exception is thrown when trying to deserialize an object with a collection field where the JSON representation contains an explicit `null`.

Conjure definition:

```
types:
  definitions:
    default-package: com.palantir.product
    objects:
      MapExample:
        fields:
          items: map<string, string>
```

JSON input:

`{"items": null}`

Exception:

```
com.fasterxml.jackson.databind.JsonMappingException: items cannot be null

	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:278)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty._throwAsIOE(SettableBeanProperty.java:611)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty._throwAsIOE(SettableBeanProperty.java:599)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty._throwAsIOE(SettableBeanProperty.java:622)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.setAndReturn(MethodProperty.java:197)
	at com.fasterxml.jackson.module.afterburner.deser.OptimizedSettableBeanProperty.setAndReturn(OptimizedSettableBeanProperty.java:111)
	at com.fasterxml.jackson.module.afterburner.deser.SettableObjectMethodProperty.setAndReturn(SettableObjectMethodProperty.java:9)
	at com.fasterxml.jackson.module.afterburner.deser.SettableObjectMethodProperty.deserializeSetAndReturn(SettableObjectMethodProperty.java:87)
	at com.fasterxml.jackson.databind.deser.BuilderBasedDeserializer.vanillaDeserialize(BuilderBasedDeserializer.java:269)
	at com.fasterxml.jackson.databind.deser.BuilderBasedDeserializer.deserialize(BuilderBasedDeserializer.java:193)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4013)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3004)
	at com.palantir.conjure.java.types.WireFormatTests.testNilMapEntryUnmarshals(WireFormatTests.java:64)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExpectedException$ExpectedExceptionStatement.evaluate(ExpectedException.java:239)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: com.palantir.logsafe.exceptions.SafeNullPointerException: items cannot be null
	at com.palantir.logsafe.Preconditions.checkNotNull(Preconditions.java:207)
	at com.palantir.product.MapExample$Builder.items(MapExample.java:108)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.setAndReturn(MethodProperty.java:193)
	... 32 more
```

## What did you want to happen?
The Conjure wire format specification explicitly states that, in this case, the object should be unmarshalled without error and the field should be initialized to be an empty collection: https://github.com/palantir/conjure/blob/master/docs/spec/wire.md#561-coercing-json-null--absent-to-conjure-types

**Comment by nmiyake** [comment]
For the collection types, the generated code is currently of the form:

```java
        @JsonSetter("items")
        public Builder items(Map<String, String> items) {
            this.items.clear();
            this.items.putAll(Preconditions.checkNotNull(items, "items cannot be null"));
            return this;
        }
```

Adding a `null` check to the JSON setter appears to resolve the issue:

```java
        @JsonSetter("items")
        public Builder items(Map<String, String> items) {
            this.items.clear();
            if (items != null) {
                this.items.putAll(items);
            }
            return this;
        }
```

Because both the `Builder` type and the `items` method are public, this would change the behavior for people who call the builder directly (calling this function with `null` input will now be equivalent to calling it with an empty map, whereas before it would fail).

**Comment by carterkozak** [comment]
I don't think that fully resolves the issue, what about an endpoint which takes a request body of `map<string,string>`? Sending null will not result in an empty collection, and I cannot imagine a case where we would want it to!
Given that, perhaps we should take a step back to understand the rationale for the quoted portion of the wire spec, and how it's supported across languages today. My assumption is that coercion from absent to empty collection is to allow certain types of API changes without breaks, but I don't think the same applies to null, so I'm curious why they're treated the same.

In the suggestion above, I would prefer if it wasn't possible to run `Foo.builder().items(null).build()` given we don't allow null anywhere else, not even for the non-optional overload of optional fields.

**Comment by carterkozak** [comment]
If we need to support null coercion, we should update the generator with something like this so we can disallow null values from the builder:
```diff
- @JsonSetter("items")
+ @JsonSetter(value = "items", nulls = Nulls.AS_EMPTY)
```

**Comment by nmiyake** [comment]
I think the *usage* of the types (broadly defined as server/client code) is independent from the spec for serialization of the objects themselves. The section of the spec I quoted deals explicitly with the latter, in the form of the expected serde behavior for fields of a given type (although even here the spec is a bit ambiguous, in that it doesn't clarify if the rules apply strictly just to these collection types, or any type that resolves to such a type -- that is, how to handle aliases of these things).

Things like request bodies are more of a concern for that part of the spec -- for example, https://github.com/palantir/conjure/blob/master/docs/spec/wire.md#23-body-parameter explicitly describes the rules around the handling of `null` for optional body parameters (it makes no mention of collection parameters -- that should probably be added, but that's a different issue having to do with the spec for clients/servers).

**Comment by nmiyake** [comment]
Thanks for the tip -- verified that the `Nulls.AS_EMPTY` seems to fix the issue, and agree that this is much cleaner from both an API and code standpoint.

Would anyone object to making this change?

If we want to have a discussion about modifying the spec I'm fine with that, but given that we publish and version the spec and the implementations, even in that scenario I feel that the right path forward is to match the current Java implementation to the current spec.

**Comment by ferozco** [comment]
Closed by #446
